Ordered, nested matching of datatypes and constants

osavaryb opened the pull request:
Added support for ordered, nested matching of datatypes and constants.

osavaryb commented:
I will add (unit)tests for this and #455 in the next few days.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 79d5ce0579cfdfe749e52e905d1d9d4c90f9cdcf
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4ec95889-b693-478d-833a-d48acf30e7a8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

osavaryb commented:
Looking into the failing tests at the moment, most of them are either changes in the number of verified goals (due to duplicating branches during CompileRBranch) or errors being thrown in different places (I will be working on reducing the number of those), but a few (e.g. Streams.dfy) are assertion failures during resolution and typechecking and I'll be looking into these this week.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: f68e9ffd49ca0e599f4c795b9b2b071df6651207
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=ea6c413b-706c-494b-9dd1-c90a8222d5be%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 3219e396ccd7ecbb47c33c12817963fce3020aa4
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=6ea64e91-d6fd-4c9b-92d0-52541db45c95%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

osavaryb commented:
Status update: at this point there are still 4 reasons why tests are failing (not counting tests in allocated):
- (8/27, e.g. dafny0/Coinductive.dfy) Changes in error messages or number of verified goals (since nested pattern match compilation may duplicate subterms)
- (3/27, e.g. dafny4/SoftwareFoundations-Basics.dfy) File type checks, but verification fails (assertion in dfy fails), probably due to unfolding bound being too low (as manually inlining constants causes verification to succeed)
- (15/27, e.g. dafny3/Paulson.dfy) Uncaught assertion failure from get_Resolved (i.e. ResolvedExpression == null) during a call to CloneNestedMatchCaseStmt, which returns an unresolved statement with unresolved subexpression. Will look more into this.
- (1/27, lambdas/StateMonad.dfy) "Error: Pattern _#Make2 is not a constructor of the given type ?" from lambdas/StateMonad.dfy. It seems like in this case the type is not known at the point where I'm compiling the nested pattern match. A likely fix for this is to add constrains to the type checking problem rather than throwing an error when checking if a pattern represent a constructor of an unresolved/unknown type.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: a08ab7af629d596f7de92e37edf5548685a35561
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=f7933733-619f-4964-af5a-494b6913a294%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 1e104d30d9c169e23ebafbcbceba78de5cd7d578
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=902bd2ea-e0c5-4c13-b612-b9689736b603%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 6405fdba44353ea492ed25721fb03a202acf8562
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=c517b45b-4b8f-4a66-9ec1-3e4b43e7a6e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 2cca090f8ef458c142897db27d36ba0dc905c463
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=c9bd5b78-82a5-4145-89d3-2c28f06b1337%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 94e9af797350adddcf5c03a5bd05ae9125f0a0ea
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=215bc6b9-a9b6-4f52-ba38-9d906d9e0acb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 4af275e82fa2ef8c10330141a2bb72ae0a688b0d
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=d4ee233e-6651-4aa9-85c5-315023c79218%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 772bebd72bd150a2902189df1f0aa951f7948aff
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=d4c18475-2763-4801-ad22-c4cb92261fb1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 038b1e18c4c63954099b6000d35e02a6995f4ec7
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=7210b2fa-22c5-4dad-9c07-c02dde8a1530%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 01cda16c40ff17bd109483f9f57b2984f1bea4f1
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0fa809f4-691a-42c2-bc52-0ff11d254595%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: c07dcac4a82d64b3eba078c34862f86a0897e784
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=46a8d6e2-ff12-46ea-a71f-35f9c09c5f4c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 0b778b9f402a8f06af58f40337049240f699aacf
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=62e5e66e-4b21-4131-a78c-b959f1382ee5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 9f6314f5d69024361090a789ff81517e6f612c52
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2838f041-e731-44f3-a79e-f84749e0dccd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 9b74275406236914d57441ad8bdfe4a782fe6a10
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2eb6ad2e-37a2-454c-bb94-966ac67719af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 04115a1a0ecb0a341b18d2a4294deb6da97f1400
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b82ff554-b817-4228-8dc9-1201f32f933d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: ba093d4a832613ea6d403c7aaba66c3886d087e6
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=32144e9b-a97a-4b4e-8b5a-b3492732be2a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 8c87357dd4c8d048033f6487ca15ac360edc6226
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=ba8f705c-b11e-4665-8a02-b758868937ff%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 894d37ea50edb3e0513d16692096600a305f9414
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=159b7b91-7f93-461a-9f28-719228ca789a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: a3698fb5cbea9a1b19138d3f9f32b12c5751c51c
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=157553d6-783c-420d-b5d7-dcc7963e4670%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 64bc6b3d931f0cfc38729fb06c8aed86d4ccd8d9
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0517d7af-2a6e-43ee-b76d-4179fbd05e0e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: c971ef150798fcc9a3effe66fe05e10833bd5400
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b63ee68b-76e0-4c20-bbf2-5925bf96e7b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: ad64854ca5bcfeb412bb82bce79f4c01e5d6b0a1
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2e6d1c9c-2a90-47ef-8585-400b4b2c334a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 0d754a498b45206b0e4288de7879979bcb0b3862
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=7f0aedfc-4944-4a04-af4f-706049f5859f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 34be7a2dbf6d28ca470258ca75ef091d729b6af4
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=6b42e3f3-b716-4bb4-aa9c-3a265fbb6628%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

RustanLeino commented on Test/dafny0/NestedPatterns.dfy.expect:
```diff
@@ -1,9 +1,9 @@
-NestedPatterns.dfy(75,2): Error: member  does not exist in datatype List
-NestedPatterns.dfy(76,2): Error: member  does not exist in datatype List
+NestedPatterns.dfy(69,2): Warning: this branch is redundant 
+NestedPatterns.dfy(75,11): Error: member _#Make1 does not exist in type List<_T0>
```
I think an attempt to use a 1-tuple should be flagged as an error already in the parser.

RustanLeino commented on Test/dafny3/Inc.dfy.expect:
```diff
@@ -8,7 +8,7 @@ Inc.dfy(83,12): Related location: This is the postcondition that might not hold.
     (0,0): anon0
     (0,0): anon3_Then
-Inc.dfy(93,2): Error BP5003: A postcondition might not hold on this return path.
+Inc.dfy(92,8): Error BP5003: A postcondition might not hold on this return path.
```
Is the more specific information about which `case` is on the failing path no longer available?

RustanLeino commented on Test/git-issues/git-issue-495.dfy.expect:
```diff
@@ -1,17 +1,17 @@
-git-issue-495.dfy(9,12): Error: value does not satisfy the subset constraints of 'nat'
+git-issue-495.dfy(7,8): Error: value does not satisfy the subset constraints of 'nat'
```
I would argue (9,12) is better error location than (7,8). Is it possible to get (9,12) instead?

osavaryb replied:
This was due to the patterns generated during pattern-match compilation possibly standing in for multiple branches (the token of the selector is used for those). I've now made reported error locations a bit more precise now by keeping track of the token of the original branch (in each RBranch), and using that in place of the selector's token when generated variables are bound in that branch.

RustanLeino commented on buildspec.yml:
```diff
@@ -12,7 +12,10 @@ phases:
       # Some recent changes have broken one of our test cases, so lock down to a specific
       # release for now. See https://github.com/dafny-lang/dafny/issues/476.
-      - git clone --branch v2.4.2 https://github.com/boogie-org/boogie.git
+      - git clone https://github.com/boogie-org/boogie.git
```
Indicate a specific branch

RustanLeino commented on Test/dafny0/ControlStructures.dfy.expect:
```diff
@@ -1,17 +1,17 @@
-ControlStructures.dfy(8,2): Error: missing case in case statement: Purple
+ControlStructures.dfy(8,8): Error: missing case in case statement: Purple
```
The new error location is okay, but I think the old one is slightly better. Is it only a matter of using a different `.tok` field in the implementation (if so, please change it) or does this involve some token that's passed in through several methods in the implementation?

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -0,0 +1,89 @@
+datatype List<T> = Nil | Cons(head: T, tail: List)
+
+// Nested Patterns
+method NestingTest (xs:List<int>) returns (r:int)
```
No space between method name and `(`

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -0,0 +1,89 @@
+
+datatype Alt = A(int) | B(int)
+datatype MyOption = Some(v: Alt) | None
+datatype MyPair = Pair(x:Alt, y:Alt)
```
According to the Dafny style guide, use a space after each `:` (here and throughout this file)

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -0,0 +1,89 @@
+    match xs
+     case Cons(y, Cons(z, zs)) => return z;
+     case Cons(y, Nil) => return y;
+     case Nil => return 0;
```
Indent with 2 spaces. Also, line up each `case` with `match` (since this `match` does not use curly braces around the `case`s).

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -0,0 +1,89 @@
+        case Cons(a, Cons(b, tl1)) => r := 0;
+        case Cons(c, tl2) => r:=1;
+        case d => r := 2;
+    }
```
Use 2 space for each indent. (Note, since this `match` uses braces around the `case`s, each `case` is indented 2 more spaces than the `match` -- as it already is, except for using 2 spaces.)

seanmcl commented:
Here's a small program that causes a crash:

```
module Result {
  datatype t<A> =
    | Success(value: A)
    | Failure(error: string)
  {
    function method flatMap<B>(f: A -> t<B>): t<B> {
      match this {
        case Success(x) => f(x)
        case Failure(y) => Failure(y)
      }
    }
  }
}

module List {
  export
    reveals t
    
  import Result

  datatype t<A> = Nil | Cons(A, t<A>)
  { 
    function method forallM(p: A -> Result.t<bool>): Result.t<bool> {
      match this {
        case Nil => Result.Success(true)
        case Cons(x, xs) => p(x).flatMap((res: bool) => match res {
          case true => xs.forallM(p)
          case false => Result.Success(false)
        })
      }
    }
  }
}
```

```
$ dafny List.dfy
Dafny 2.3.0.10506
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.Dafny.Resolver.MakeIfFromContainers (Microsoft.Dafny.Resolver+MatchTempInfo mti, Microsoft.Dafny.Expression matchee, System.Collections.Generic.List`1[T] blocks, Microsoft.Dafny.Resolver+SyntaxContainer def) [0x000ac] in <2435363751254b359fdacc4663832a39>:0
  at Microsoft.Dafny.Resolver.MakeIfFromContainers (Microsoft.Dafny.Resolver+MatchTempInfo mti, Microsoft.Dafny.Expression matchee, System.Collections.Generic.List`1[T] blocks, Microsoft.Dafny.Resolver+SyntaxContainer def) [0x00142] in <2435363751254b359fdacc4663832a39>:0
  at Microsoft.Dafny.Resolver.MakeIfFromContainers (Microsoft.Dafny.Resolver+MatchTempInfo mti, Microsoft.Dafny.Expression matchee, System.Collections.Generic.List`1[T] blocks, Microsoft.Dafny.Resolver+SyntaxContainer def) [0x00142] in <2435363751254b359fdacc4663832a39>:0
  at Microsoft.Dafny.Resolver.CompileRBranch (Microsoft.Dafny.Resolver+MatchTempInfo mti, System.Collections.Generic.List`1[T] matchees, System.Collections.Generic.List`1[T] branches, Microsoft.Dafny.ICodeContext codeContext) [0x00f48] in <2435363751254b359fdacc4663832a39>:0
  at Microsoft.Dafny.Resolver.CompileNestedMatchExpr (Microsoft.Dafny.NestedMatchExpr e, Microsoft.Dafny.ICodeContext codeContext) [0x000c9] in <2435363751254b359fdacc4663832a39>:0
  at Microsoft.Dafny.Resolver.ResolveNestedMatchExpr (Microsoft.Dafny.NestedMatchExpr me, Microsoft.Dafny.Resolver+ResolveOpts opts) [0x00252] in <2435363751254b359fdacc4663832a39>:0
...
```

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -0,0 +1,89 @@
+
+/*
+*  This file contains a collection of tests for features modified or introduced in PR 458
+*/
```
Also add tests where the patterns use literals of various types: `bool`, `int`, some `newtype`, `real`, `char`, `string`. Are sequence display expressions (e.g., `[5, _, x]`) allowed? If so, test them, too. I'm guessing that display expressions for sets, multisets, and maps are not allowed (since their components are not ordered).

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -423,6 +423,7 @@ enum ValuetypeVariety { Bool = 0, Int, Real, BigOrdinal, Bitvector, Map, IMap, N
           ResolveModuleExport(literalDecl, sig);
           var good = ResolveModuleDefinition(m, sig);
 
+
```
remove this blank line

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -5503,6 +5507,7 @@ public CheckTypeInference_Visitor(Resolver resolver, ICodeContext codeContext)
         }
       }
       protected override void VisitOneExpr(Expression expr) {
+
```
remove this blank line

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -5619,6 +5624,7 @@ public CheckTypeInference_Visitor(Resolver resolver, ICodeContext codeContext)
           CheckTypeIsDetermined(expr.tok, expr.Type, "variable");
         } else if (CheckTypeIsDetermined(expr.tok, expr.Type, "expression")) {
           if (expr is BinaryExpr) {
+
```
remove this blank line

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -3075,6 +3076,9 @@ public class ModuleBindings
         }
+      } else if (expr is NestedMatchExpr) {
+        var e = (NestedMatchExpr)expr;
+        CheckDestructsAreAbstemiousCompliant(e.ResolvedExpression);
```
Are there some test cases for this?

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -6774,6 +6786,7 @@ class GhostInterest_Visitor
       /// </summary>
       public void Visit(Statement stmt, bool mustBeErasable) {
         Contract.Requires(stmt != null);
+//        Console.WriteLine("DEBUG: visit pre assume: isghost:{0}, mustbeerasable:{1}, statement:{2}", codeContext.IsGhost, mustBeErasable, stmt.ToString());
```
remove

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+
+  public IToken[] BranchTok;
+
+    public int[] BranchIDCount; // Records the number of copies of each branch
```
indentation

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+public abstract class SyntaxContainer
+{}
+
+public class CExpr:SyntaxContainer
```
spaces around the `:` (here and elsewhere)

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+/* Temporary information about the Match being desugared  */
+public class MatchTempInfo{
+  public IToken Tok;
+
```
This is still in the lexical scope of the previous class, right? If so, these things should be indented.

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+}
+
+
+
```
Remove repeated empty lines (here and elsewhere)

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
-          } else {
-            ctorsList.Add(new Dictionary<string, DatatypeCtor>());
+    if (branch is RBranchStmt) {
+      if(substitute){
```
spaces

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+        foreach (Statement ss in ((RBranchStmt)branch).Body) {
+          Statement clone = cloner.CloneStmt(ss);
+          list.Add(clone);
           }
```
unindent

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+        MatchCaseExprSubstituteCloner cloner = new MatchCaseExprSubstituteCloner(new BoundVar(expr.tok, name, type), oldVar);
+        ((RBranchExpr)branch).Body = cloner.CloneExpr(((RBranchExpr)branch).Body);
+      } else {
+      var cBVar = new BoundVar(expr.tok, name, type);
```
Please check all formatting before doing the PR. I will stop commenting on code style in this file.

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -2579,46 +2579,78 @@ ExistentialGuard<out Expression e, bool allowLambda>
+      [ ExtendedPattern<out pat>          (. arguments.Add(pat); .)
+      { "," ExtendedPattern<out pat>      (. arguments.Add(pat); .)
+      }]
+    ")"                                   (. id = t;
```
Wouldn't it be better to use the token from the `(` instead of the token for the `)`?

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -2579,46 +2579,78 @@ ExistentialGuard<out Expression e, bool allowLambda>
+      }]
+    ")"                                   (. id = t;
+                                             // make sure the tuple type exists
+                                             theBuiltIns.TupleType(id, arguments.Count, true);
```
Report an error here in the parser if the input tried to use a 1-tuple. Dafny does not support 1-tuples, because they are considered to be just parentheses around an expression. A pattern is not allowed to use parentheses, so therefore this should be flagged as an error here.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 35c4aaf4dfed16df421f0434d6610053bb13e1c2
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=da535d8b-e2b2-478b-8ca4-cbc48847ec24%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: c8afb4bacfb149a6b1d839db71eef66301d5516e
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=15c8fbe7-07b3-49d1-be94-7be1bea32cb3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: e18ba26504ec609e0d68b3b17da0676786ece5ae
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=1641bb0b-da1e-4b3a-b986-937ee2002500%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

seanmcl commented:
I'm puzzled by one aspect of pattern matching, though it is present in mainline as well.  Say I have the following:

```
module List {
  datatype t<A> = Nil | Cons(a: A)
}
```

Then in every other functional language I know of I would match like so:

```
import List
match (foo: List.t<int>) {
  List.Nil => 5
  List.Cons(x, y) => 6
}
```

But in Dafny I'm not allowed to qualify the module name on the constructor.  I need to do

```
import List
match (foo: List.t<int>) {
  Nil => 5
  Cons(x, y) => 6
}
```

Since I don't import List as opened, I can't use the values `Cons` or `Nil` in any other context without the module qualifier.  I understand that Dafny can figure out based on type information what the legal constructor names are, but just as a person reading the code, it's weird to see unqualified names that I don't explicitly import.  Java is the only language I'm aware of that has this property on `switch` statements, and I've always thought that was weird also, so maybe it's just an aesthetic judgment.  

Since you're now well versed in the code, how hard do you think it would be to allow module qualifiers on patterns?

osavaryb commented on Docs/DafnyRef/DafnyRef.mdk:
```diff
@@ -6865,41 +6864,50 @@ var k := 10 / x; // error, may divide by 0.
+  | "(" [ ExtendedPattern { "," ExtendedPattern } ] ")"
   )
 
 CasePattern =
```
Yes, CasePattern are still used as the lhs of LetExpr. I could mention this and/or move the CasePattern discussion to the Let Expression section

osavaryb commented on Docs/DafnyRef/DafnyRef.mdk:
```diff
@@ -5829,22 +5829,21 @@ MatchStmt = "match" Expression(allowLemma: true, allowLambda: true)
-tuples. The expression after the `match` keyword is the (co)inductive
-value being matched. The expression is evaluated and then matched against
-each of the case clauses.
+The `match` statement is used to do case analysis on a value of inductive,  co-inductive or base type, or of a tuple. The expression after the `match` keyword is called the _selector_. The expression is evaluated and then matched against
```
yes, it can be a newtype, I'll add an example to the Test file (and update DafnyRef)!

osavaryb commented on Source/Dafny/Dafny.atg:
```diff
@@ -2116,7 +2116,7 @@ OneStmt<out Statement/*!*/ s>
   | IfStmt<out s>
   | WhileStmt<out s>
-  | MatchStmt<out s>
+  | NestedMatchStmt<out s>
```
There are no more rules producing MatchStmt in the Parser (and pre-resolution), but MatchStmt is still used post-resolution as ordered and disjoint pattern matching produced by pattern-match compilation. At the moment both MatchStmt and NestedMatchStmt exist in DafnyAST, but we could merge them...

seanmcl commented:
There is a performance problem with the compiler, possibly due to this change.  I have code with the following shape

```
match E1 { 
  case C1 => match E2 {
    case C2 => match E3 {
      case C3 => match E4 {
        case D1 => ...
        case D2 => ...
        ...
        case D13 => ...
      }
    }
  }
    
}
```

which is timing out the compiler.  When I comment out D4-D13, compiler takes ~80s.  Commenting out D5-D13 compiler takes ~120s.  So there's possibly some non-linear behavior here.  There is no problem with the verifier for this code.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 39fa22b39f1ca97c77b7b8bcc3f747b67f2c4b46
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=eb32d0ef-62c9-4f28-a63e-bce9aff33c68%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

septract commented:
> I'm inclined to just merge this when Rustan is happy with the shape, then fix this in another PR. What do you think?

Serdar is pushing us to close, so I think fixing this would have to be a follow on project at this point.

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 42f2d2cf535fdb09f83ede5efa8a9af0116cc17d
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fa0d969b-4e1a-43c8-ad5c-32032e595674%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 83016c3b52d1c7899356ff602f6060d61822ad1a
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=5d2fa96f-2a56-4d9e-ab6a-0b6d060201af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

osavaryb commented on Test/allocated1/dafny0/NatTypes.dfy.expect:
```diff
@@ -31,10 +31,9 @@ Execution trace:
     (0,0): anon9_Then
     (0,0): anon13_Then
-NatTypes.dfy(91,15): Error: assertion violation
+NatTypes.dfy(87,20): Error: value does not satisfy the subset constraints of 'nat'
```
This is fixed in 42f2d2c (CompileRBranch [3 and 3.3] was introducing subtyping constraints in the wrong direction).

osavaryb commented on Test/dafny0/ModuleExport.dfy.expect:
```diff
@@ -5,8 +5,7 @@ ModuleExport.dfy(32,7): Error: expected method call, found expression
 ModuleExport.dfy(53,9): Error: unresolved identifier: k
-ModuleExport.dfy(66,2): Error: Type mismatch: expected constructor of type Data.  Got Orang.
-ModuleExport.dfy(66,2): Error: member Orang does not exist in datatype Data
+ModuleExport.dfy(66,2): Warning: this branch is redundant 
```
Since we allow for variables as patterns, Orang might be a variable of type Data so we won't be able to disallow this

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: b9c46f74a5aed0fc3d835102065f076c67124efb
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4799c656-7a0c-4efb-8d5d-4ba261d9c57a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 4ad09bc90127547834a12a5f9b1d9592a1435c5b
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=67fba8d8-5c74-418f-bbea-f1d3fbad215b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 28dfdd6730643ed512d0c1761c552ec2c9cd4506
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a15f7a80-1bf7-4339-90e2-50b1536062ed%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

RustanLeino commented:
Thank you, the stylized code is easier to read. Here are various comments to address or discuss.

## Sean's comments

Yes, not allowing qualified names in `case`s is weird. Can you please file an Issue, @seanmcl ? Let's not address it in this PR. @osavaryb , do you have a feel for how much effort this would be to do?

I agree about the performance issue--it needs to be addressed, but let's do that after this PR goes in.

## subtype/supertype NatTypes

I'm really happy that the issue in `NatTypes.dfy` was fixed.

## Return path, inductive lemmas

In `Test/dafny3/Inc.dfy.expect`, the "this return path" location has changed (from before the PR). All the places where this occurs (which I think is just in this file) in this file seem to be in `inductive lemma`s. Is there a bug to look into there?

## Location of case, not variable declaration

L9 of `Test/git-issues/git-issue-495.dfy` L9 and L565 of `Test/dafny0/ResolutionErrors.dfy` seem like they may be the same problem. It seems that the bound variable gets the location of the enclosing constructor. Can this be changed?

## Improved locations

The error locations in `Test/dafny4/git-issue281.dfy` are a nice improvement over what was there before.

## Missing-case error messages

In a program like:

    function F(xs: List): int {
      match xs  // missing case in case statement: Nil
      case Cons(x, Cons(y, rest)) => 5
      case Nil => 8
    }

I don't find the error message as helpful as I would wish. The missing `Nil` is not the one that is given in the function, but the `Nil` in `Cons(_, Nil)`. Is there a way to make such error messages more specific?

There's also something else with this message. There are no "case statements" in Dafny--they are "`match` statements". Furthermore, the function above does not have a "`match` statement" but a "`match` expression".

## Redundancy checking

I'm finding that the feature of redundancy checking is really nice, when the nesting of patterns gets complicated and it's easy to get confused.

## Verification message for missing case

In the program

    datatype Dt = Real(real)
    method U(d: Dt) {
      match d  // assertion violation
      case Real(15.0) =>
    }

This program is erroneous, because it's missing cases for non-15.0 reals. However, the error message is just the mysterious "assertion violation". This should be improved.

## Checking redundant cases

The analysis incorrectly reports a "redundant branch" error for this program:

    datatype PairOfNumbers = P(int, int)
    method M(p: PairOfNumbers) {
      match p
      case P(10, 10) =>
      case P(10, 11) =>  // this branch is redundant
      case _ =>
    }

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 147b533c62a68827c34b5916b2fcffc7f08f995a
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=1235a095-fab0-41f3-9006-dd009779ab89%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

robin-aws commented:
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: a9ec0cd169d59fb7d49fdce64ab23f3f4221d195
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fea3366d-26ec-410a-b262-31d77cc76d2d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

osavaryb commented:
> ## Return path, inductive lemmas
> In `Test/dafny3/Inc.dfy.expect`, the "this return path" location has changed (from before the PR). All the places where this occurs (which I think is just in this file) in this file seem to be in `inductive lemma`s. Is there a bug to look into there?
> ## Location of case, not variable declaration> L9 of `Test/git-issues/git-issue-495.dfy` L9 and L565 of `Test/dafny0/ResolutionErrors.dfy` seem like they may be the same problem. It seems that the bound variable gets the location of the enclosing constructor. Can this be changed?

I'll have a closer look at this over the weekend, the second issue (with git-issue-495 and ResolutionErrors) is due to how the selector's token (used when generating fresh variables in the PM compiler) is overwritten by the RBranch's token when let-bound in that branch. I should be able to get better error location by overwriting these token earlier. 

> ## Missing-case error messages
> In a program like:
> 
> ```
> function F(xs: List): int {
>   match xs  // missing case in case statement: Nil
>   case Cons(x, Cons(y, rest)) => 5
>   case Nil => 8
> }
> ```
> 
> I don't find the error message as helpful as I would wish. The missing `Nil` is not the one that is given in the function, but the `Nil` in `Cons(_, Nil)`. Is there a way to make such error messages more specific?
This is a bit complicated since the missing case error is thrown by the verifier while doing a semantic exhaustiveness check...I may be able to generate the missingCases during NestedMatch desugaring instead of while resolving the Match.

> There's also something else with this message. There are no "case statements" in Dafny--they are "`match` statements". Furthermore, the function above does not have a "`match` statement" but a "`match` expression".
> 
This is unrelated to the PR, but I will update the error messages in Translator.cs

> ## Verification message for missing case
> In the program
> 
> ```
> datatype Dt = Real(real)
> method U(d: Dt) {
>   match d  // assertion violation
>   case Real(15.0) =>
> }
> ```
> 
> This program is erroneous, because it's missing cases for non-15.0 reals. However, the error message is just the mysterious "assertion violation". This should be improved.
> This example gets desugared to  match d {
    case Real(_mcc#0: real) =>
      {
        assert _mcc#0 == 15.0;
      }
I may be able to do better by attaching a custom error message to the assertion generated by the lack of default case in a constant match.

> ## Checking redundant cases
> The analysis incorrectly reports a "redundant branch" error for this program:
> 
> ```
> datatype PairOfNumbers = P(int, int)
> method M(p: PairOfNumbers) {
>   match p
>   case P(10, 10) =>
>   case P(10, 11) =>  // this branch is redundant
>   case _ =>
> }
> ```
Fixed (and added as test case in PatternMatching.dfy, thank you!

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -131,6 +131,14 @@ method NewTypeTest() returns (r: int) {
+  case P(11, a) => r := 1;
+  case P(10, 11) => r := 2;
+  case _ => r := 3;
+}
```
Good. This tests that no "redundant case" error is generated.

To additionally test that the verifier behaves appropriately on this example, add some postconditions (phrased in some way that doesn't use `match`). For example, add destructors to `PairOfNumbers`, like:

    datatype PairOfNumbers = P(0: int, 1: int)

and then add

    ensures p.0 == 11 && p.1 == 10 ==> r == 0
    ensures p.0 == 11 && p.1 != 10 ==> r == 1
    ensures p.0 == 11 || p.1 == 11 || r == 3

Another way to test the verifier would be to add an `assert a != 10;` on L138 or to add

    assert p.1 == 11 ==> p.0 != 10 && p.1 != 11;

to L140.

Then, to test the compiler, add a couple of calls to `M` from `Main`.

Once you have the interesting test method (here, `M`), it's easy to add these additional tests, and it's good to do so habitually.

RustanLeino commented on Source/Dafny/Translator.cs:
```diff
@@ -7673,7 +7673,7 @@ private class WFOptions
             builder.Add(new Bpl.HavocCmd(me.tok, havocIds));
           }
-          b.Add(Assert(me.tok, Bpl.Expr.False, "missing case in case statement: " + missingCtor.Name));
+          b.Add(Assert(me.tok, Bpl.Expr.False, "missing case in match expression: " + missingCtor.Name));
```
Thanks for making this change. Was no test case affected? If not, please add some test case that produces this error message.

RustanLeino replied:
Oh, the next commit did just that.

RustanLeino commented on Test/patterns/PatternMatching.dfy:
```diff
@@ -149,6 +149,31 @@ method M(p: PairOfNumbers) returns (r:int)
+  match x {
+    case A(i) =>
+      match y {
+        case Branch(_, _, Branch(_, b, _)) =>
+	  match (b, z) {
```
indentation

RustanLeino commented on Test/patterns/PatternMatching.dfy.expect:
```diff
@@ -6,4 +6,4 @@ NestedVariableTest([2::3::4::5::6]) = 0, should return 0
 M(P(11,10)) = 0, should return 0 
 M(P(-1,10)) = 3, should return 3 
-Testing MultipleNestedMatch: 012345, should return 012345
+Testing MultipleNestedMatch: 012345, should return 012345 
```
If you can easily avoid having a space at the end of a line in the test output, then that would be good. Certain editors or commit filters may otherwise erase the space.
